pkg/serviceaccounts/controllers/docker_registry_service.go

```diff
@@ -0,0 +1,203 @@
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/credentialprovider"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+	"github.com/golang/glog"
+)
+
+// DockerRegistryServiceControllerOptions contains options for the DockerRegistryServiceController
+type DockerRegistryServiceControllerOptions struct {
+	// Resync is the time.Duration at which to fully re-list service accounts.
+	// If zero, re-list will be delayed as long as possible
+	Resync time.Duration
+
+	RegistryNamespace string
+
+	RegistryServiceName string
+
+	DockercfgController *DockercfgController
+}
+
+// NewDockerRegistryServiceController returns a new *DockerRegistryServiceController.
+func NewDockerRegistryServiceController(cl client.Interface, options DockerRegistryServiceControllerOptions) *DockerRegistryServiceController {
+	e := &DockerRegistryServiceController{
+		client: cl,
+	}
+
+	_, e.serviceController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return e.client.Services(options.RegistryNamespace).List(labels.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return e.client.Services(options.RegistryNamespace).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Service{},
+		options.Resync,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    e.serviceAdded,
+			UpdateFunc: e.serviceUpdated,
+			DeleteFunc: e.serviceDeleted,
+		},
+	)
+	e.registryServiceName = options.RegistryServiceName
+	e.dockercfgController = options.DockercfgController
+
+	return e
+}
+
+// DockerRegistryServiceController manages ServiceToken secrets for Service objects
+type DockerRegistryServiceController struct {
+	stopChan chan struct{}
+
+	client client.Interface
+
+	registryServiceName string
+
+	dockercfgController *DockercfgController
+
+	serviceController *framework.Controller
+}
+
+// Runs controller loops and returns immediately
+func (e *DockerRegistryServiceController) Run() {
+	if e.stopChan == nil {
+		e.stopChan = make(chan struct{})
+		go e.serviceController.Run(e.stopChan)
+	}
+}
+
+// Stop gracefully shuts down this controller
+func (e *DockerRegistryServiceController) Stop() {
+	if e.stopChan != nil {
+		close(e.stopChan)
+		e.stopChan = nil
+	}
+}
+
+// serviceAdded reacts to the creation of a docker-registry service by updating all service account dockercfg secrets and
+// changing all interestedURLs
+func (e *DockerRegistryServiceController) serviceAdded(obj interface{}) {
+	service := obj.(*api.Service)
+	if service.Name != e.registryServiceName {
+		return
+	}
+
+	serviceLocation := service.Spec.PortalIP
```

Comment:
@ncdc same question here, do we need to include a port from the service's spec.Ports in the .dockercfg registry key? If there are multiple ports, how should we identify the right one?